src/crypto/transform.py

```diff
@@ -90,40 +79,130 @@ def date_dim_row(self) -> list:
             month_end=(d + relativedelta(day=31)).strftime("%Y-%m-%d")
         )._asdict()
         return [row]
-    
-    def _get_name_dim_keys(self):
-        query = """SELECT DISTINCT symbol FROM {name_dim}""".format(name_dim=NAME_DIM)
-        job = self.bq_client.query(query=query)
-        result = job.result()
-        return result
 
-    def name_dim_rows(self, crypto_data: dict) -> None:
-        result = self._get_name_dim_keys()
-        symbols = [row["symbol"] for row in result]
+    def name_dim_rows(self, crypto_data: dict) -> list:
+        existing_symbols_query = """SELECT symbol FROM {name_dim}""".format(name_dim=Table.NAME_DIM.value)
+        result = self.bq_client.query(query=existing_symbols_query).result()
+        existing_symbols = set(row["symbol"] for row in result)
 
         rows = []
         for row in crypto_data:
-            if row["symbol"] in symbols:
+            if row["symbol"] in existing_symbols:
                 continue
             name_dim = NameDim(name_key=row["name"], symbol=row["symbol"], slug=row["slug"])
             rows.append(name_dim._asdict())
         return rows
 
+    def _get_tags(self, crypto_data: dict, existing_tags: set) -> set:
+        tags = set()
+        for row in crypto_data:
+            for tag in row["tags"]:
+                if tag in existing_tags:
+                    continue
+                tags.add(tag)
+        return tags
+    
+    def tag_dim_rows(self, crypto_data: dict) -> list:
+        existing_tags_query = """SELECT tag FROM {tag_dim}""".format(tag_dim=Table.TAG_DIM.value)
+        result = self.bq_client.query(query=existing_tags_query).result()
+        existing_tags = set(row["tag"] for row in result)
+
+        tags = self._get_tags(crypto_data=crypto_data, existing_tags=existing_tags)
+        
+        max_key = 0
+        if tags and not Project.INITIAL_LOAD.value:
+            tag_query = """SELECT MAX(tag_key) AS max_key FROM {tag_dim}""".format(tag_dim=Table.TAG_DIM.value)
+            result = self.bq_client.query(query=tag_query).result()
+            max_key = [row["max_key"] for row in result][0]
+
+        rows = []
+        start = max_key if max_key else 1        
+        for key, tag in enumerate(tags, start=start):
+            tag_dim = TagDim(tag_key=key, tag=tag)
+            rows.append(tag_dim._asdict())
+        return rows
+
+    def name_tag_bridge_table(self, date_key: date, crypto_data: dict) -> list:
+        tag_query = """SELECT tag_key, tag FROM {tag_dim}""".format(tag_dim=Table.TAG_DIM.value)
+        result = self.bq_client.query(query=tag_query).result()
+        tag_dim = {row["tag_key"]: row["tag"] for row in result}
+
+        rows = []     
+        for row in crypto_data:
+            name = row["name"]
+            tags = row["tags"]
+            if tags:
```

Comment:
You could also inverse and continue:

```
if not tags:
    continue
```

Then you win one indent back for the for loop next.

Reply:
Dang, this is great

Reply:
Thanks. Flat is better than nested, see https://www.linkedin.com/feed/update/urn:li:activity:7084539098790547456/